### ISSUE TITLE
tenant cluster nodes never recover after KubeVirt VM restarts

### ISSUE THREAD

**Issue by davidvossel** [kubernetes-sigs/cluster-api-provider-kubevirt#70]
/kind bug

If a KubeVirt VM restarts which is hosting a tenant cluster node, that node never comes back online within the tenant cluster.

Steps to reproduce.
1. Create a tenant cluster with capk
2. restart a tenant cluster VM `virtctl restart <vm>`
3. That node never rejoins the tenant cluster despite the VM recovering 

At step 3, KubeVirt will bring the VM back online, but internally that VM will never rejoin the tenant cluster for a few reasons.
1. The VM in our examples uses the ephemeral `containerDisk` volume type, which is not persistent across restarts. This means that the node would have to re-join the kubeadm cluster, but it can't because the token from the initial bootstrap expires.
2. The VM's IP address changes after restart, but the hostname renames the same. It's unclear if this inconsistency is tolerated when re-joining the cluster after a restart.

Here's the error that occurs on VM boot after restart with the kubeadm bootstrapper.
```error execution phase preflight: couldn't validate the identity of the API Server: could not find a JWS signature in the cluster-info ConfigMap for token ID "uavkdm"```

The tenant cluster says that the node is `NotReady`

```
# kubectl get nodes
NAME                            STATUS     ROLES                  AGE   VERSION
kvcluster-control-plane-gsxzx   Ready      control-plane,master   41m   v1.21.0
kvcluster-md-0-k5srs            NotReady   <none>                 39m   v1.21.0
```

However, both the Machine and KubeVirtMachine object indicate that the Node is still available, which seems wrong.
```
$ ./kubevirtci kubectl get machine
selecting docker as container runtime
NAME                              CLUSTER     NODENAME                        PROVIDERID                                 PHASE     AGE   VERSION
kvcluster-control-plane-9hlg7     kvcluster   kvcluster-control-plane-gsxzx   kubevirt://kvcluster-control-plane-gsxzx   Running   70m   v1.21.0
kvcluster-md-0-59885c486b-w5wfx   kvcluster   kvcluster-md-0-k5srs            kubevirt://kvcluster-md-0-k5srs            Running   70m   v1.21.0
```

machine status
```
status:
  addresses:
  - address: kvcluster-md-0-k5srs
    type: Hostname
  - address: 10.244.140.77
    type: InternalIP
  - address: 10.244.140.77
    type: ExternalIP
  - address: kvcluster-md-0-k5srs
    type: InternalDNS
  bootstrapReady: true
  conditions:
  - lastTransitionTime: "2022-01-14T19:23:33Z"
    status: "True"
    type: Ready
  - lastTransitionTime: "2022-01-14T19:21:18Z"
    status: "True"
    type: BootstrapReady
  - lastTransitionTime: "2022-01-14T19:23:33Z"
    status: "True"
    type: InfrastructureReady
  - lastTransitionTime: "2022-01-14T19:57:36Z"
    message: 'Node condition MemoryPressure is Unknown. Node condition DiskPressure
      is Unknown. Node condition PIDPressure is Unknown. Node condition Ready is Unknown. '
    reason: NodeConditionsFailed
    status: Unknown
    type: NodeHealthy
  infrastructureReady: true
  lastUpdated: "2022-01-14T19:23:33Z"
  nodeInfo:
    architecture: amd64
    bootID: 697101ac-583b-44e1-8bf4-3535f70bbefa
    containerRuntimeVersion: cri-o://1.21.3
    kernelVersion: 5.14.10-300.fc35.x86_64
    kubeProxyVersion: v1.21.0
    kubeletVersion: v1.21.0
    machineID: 3f37466c56de5676a274894afc91998f
    operatingSystem: linux
    osImage: Fedora Linux 35 (Cloud Edition)
    systemUUID: 3f37466c-56de-5676-a274-894afc91998f
  nodeRef:
    apiVersion: v1
    kind: Node
    name: kvcluster-md-0-k5srs
    uid: 9f70c294-e77e-4539-aa86-e8d6fbd92279
  observedGeneration: 3
  phase: Running
```

and kubevirt machine status
```
status:
  addresses:
  - address: kvcluster-md-0-k5srs
    type: Hostname
  - address: 10.244.140.77
    type: InternalIP
  - address: 10.244.140.77
    type: ExternalIP
  - address: kvcluster-md-0-k5srs
    type: InternalDNS
  conditions:
  - lastTransitionTime: "2022-01-14T19:23:33Z"
    status: "True"
    type: Ready
  - lastTransitionTime: "2022-01-14T19:23:33Z"
    status: "True"
    type: BootstrapExecSucceeded
  - lastTransitionTime: "2022-01-14T19:23:02Z"
    status: "True"
    type: VMProvisioned
  nodeupdated: true
  ready: true
```

**Comment by davidvossel** [comment]
I think in order to solve this, we shouldn't allow ephemeral VMs using a container disk to restart. Instead I think the entire machine needs to be replaced if the ephemeral VM ever goes down.

**Comment by davidvossel** [comment]
I tested this using a VM with persistent storage and the node survives the restart with the kubeadm provider. The IP address changes, but that IP is updated in the node's entry in the tenant cluster and everything seems to continue to work as expected.

Below is an example of the KubeVirtMachine with persistent storage. This spec requires the containerized-data-importer (cdi) to be installed on the management cluster in order to import the containerDisk into a PVC.

```
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha1
kind: KubevirtMachine
metadata:
  annotations:
    cluster.x-k8s.io/cloned-from-groupkind: KubevirtMachineTemplate.infrastructure.cluster.x-k8s.io
    cluster.x-k8s.io/cloned-from-name: kvcluster-md-0
  creationTimestamp: "2022-01-14T20:44:28Z"
  finalizers:
  - kubevirtmachine.infrastructure.cluster.x-k8s.io
  generation: 2
  labels:
    cluster.x-k8s.io/cluster-name: kvcluster
    cluster.x-k8s.io/deployment-name: kvcluster-md-0
    machine-template-hash: "1544170426"
  name: kvcluster-md-0-z64nk
  namespace: default
  ownerReferences:
  - apiVersion: cluster.x-k8s.io/v1beta1
    blockOwnerDeletion: true
    controller: true
    kind: Machine
    name: kvcluster-md-0-59885c486b-khqjs
    uid: a3736476-83c0-410a-aa13-aef8b700749d
  resourceVersion: "19472"
  uid: c58bae24-b6cf-4d1c-807c-9d1a860250cf
spec:
  providerID: kubevirt://kvcluster-md-0-z64nk
  virtualMachineTemplate:
    spec:
      dataVolumeTemplates:
      - metadata:
          name: fedora-dv
        spec:
          pvc:
            accessModes:
            - ReadWriteOnce
            resources:
              requests:
                storage: 5Gi
            storageClassName: local
          source:
            http:
              url: docker://quay.io/kubevirtci/fedora-kubeadm:35
      runStrategy: Always
      template:
        spec:
          domain:
            cpu:
              cores: 2
            devices:
              disks:
              - disk:
                  bus: virtio
                name: dv-volume
            memory:
              guest: 4Gi
          volumes:
          - dataVolume:
              name: fedora-dv
            name: dv-volume
```

**Comment by davidvossel** [comment]
/close

Works now because we have the ability to use DataVolumeTemplates within the KubeVirtMachine to create persistent nodes.

**Comment by adkafka** [comment]
I came across this issue because I was seeing the same behavior still. In hopes that we can avoid the added complexity of persistence and with a desire to still treat our servers as cattle, I found a different solution, along the same vein as what @davidvossel suggested:

> I think in order to solve this, we shouldn't allow ephemeral VMs using a container disk to restart. Instead I think the entire machine needs to be replaced if the ephemeral VM ever goes down.

We can set `spec.runStrategy: Once` on the `VirtualMachine` / `KubevirtMachineTemplate` spec, and add a [`MachineHealthChecks`](https://cluster-api.sigs.k8s.io/tasks/automated-machine-management/healthchecking) so Cluster API will replace an unhealthy node. This leads to the desirable behavior when a virt-launcher pod is terminated (at least from my perspective).

Leaving this comment here in case other folks run into this as well.
